netket/stats/mc_stats.py

```diff
@@ -98,42 +101,87 @@ def __getattr__(self, name):
         elif name in ("variance", "Variance"):
             return self.variance
         elif name in ("error_of_mean", "Sigma"):
-            return self.error_of_mean
+            return self.error
         elif name in ("R_hat", "R"):
-            return self.R_hat
+            return self.r_hat
         elif name in ("tau_corr", "TauCorr"):
             return self.tau_corr
         else:
             raise AttributeError(
                 "'Stats' object object has no attribute '{}'".format(name)
             )
 
+    @property
+    def n_chains(self):
+        return self._μ.size * mpi.n_nodes
 
-def _get_blocks(data, block_size):
-    chain_length = data.shape[1]
+    @struct.property_cached(pytree_node=True)
+    def mean(self) -> float:
+        return _mean(self._μ)
 
-    n_blocks = int(np.floor(chain_length / float(block_size)))
+    @struct.property_cached(pytree_node=True)
+    def variance(self) -> float:
+        return _mean(self._μ2) - self.mean ** 2
 
-    return data[:, 0 : n_blocks * block_size].reshape((-1, block_size)).mean(axis=1)
+    @struct.property_cached(pytree_node=True)
+    def error(self) -> float:
+        return jnp.sqrt(self.batch_variance / self.n_chains)
 
+    @struct.property_cached(pytree_node=True)
+    def r_hat(self) -> float:
+        if self.n_chains > 1:
+            return jnp.sqrt(
+                (self._n - 1) / self._n + self.batch_variance / self.variance
+            )
+        else:
+            return jnp.nan
+
+    @struct.property_cached(pytree_node=True)
+    def tau_corr(self) -> float:
+        return jnp.clip(0.5 * (self._n * self.batch_variance / self.variance - 1), 0)
+
+    @struct.property_cached(pytree_node=True)
+    def batch_variance(self) -> float:
+        return _var(self._μ)
+
+    @struct.property_cached(pytree_node=True)
+    def block_variance(self) -> float:
+        return _mean(self._μ2_w) - self.mean ** 2
+
+    @property
+    def block_size(self):
+        return self._w
+
+    def merge(self, other):
+        """
+        Merge the two statistics.
+        """
+        if self.n_chains != other.n_chains:
+            raise ValueError(
+                "Cannot sum two Stats objects for a different number of chains."
+            )
 
-def _block_variance(data, l):
-    blocks = _get_blocks(data, l)
-    ts = _total_size(blocks)
-    if ts > 0:
-        return _var(blocks), ts
-    else:
-        return jnp.nan, 0
+        n = self._n + other._n
 
+        a1 = self._n / n
+        a2 = other._n / n
 
-def _batch_variance(data):
-    b_means = data.mean(axis=1)
-    ts = _total_size(b_means)
-    return _var(b_means), ts
+        μ = a1 * self._μ + a2 * other._μ
+        μ2 = a1 * self._μ2 + a2 * other._μ2
```

Comment:
this might be numerically unstable? I would suggest to implement this https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance (the parallel version) 
```python
def parallel_variance(n_a, avg_a, M2_a, n_b, avg_b, M2_b):
    n = n_a + n_b
    delta = avg_b - avg_a
    M2 = M2_a + M2_b + delta ** 2 * n_a * n_b / n
    var_ab = M2 / (n - 1)
    return var_ab
```

I think this is what we used to have in the C++ version

Reply:
the code you are referring to here is to `merge` two Stats objects in order to obtain a new `Stats` object
The code for `parallel_varance` you posted here is to compute the variance given two `Stats` objects.

Moreover, I expect the two stats object to have comparable `n` (because they refer to two minibatches) so I do think that it should be numerically stable...

Why do you think summing the means that way is unstable?

Reply:
https://github.com/netket/netket/blob/v2.0.1/Sources/Stats/onlinestat.hpp#L66  that's how I was merging bins

Reply:
But here you are storing the variance, I am storing only the mean of the square values.

Are there arguments suggesting the latter is more stable? 

also, your algorithm is a more traditional _online_ implementation where at every iteration you add one entry to your accumulator.

my implementation does not do this (bough in principle you can). I had tried and benchmarked and due to Jax this would me extremely slow. We need something where we can batch the operations.

instead my implementation computes some things on mini batches and then you merge the mini batches.

Reply:
Ah, no, ok.

I agree that accumulating the sum of squares is less stable.

i think I can use the parallel merge of the algorithm you linked to... bough I suspect it will be slower.

do we really need all that precision when estimating the error?

Reply:
Indeed, now you are computing the variance as the difference of two means, essentially, this is known to be (extremely?) unstable numerically, so I wouldn't do it. Instead of the means of the square you should really store the variance directly and update it when merging!